provider/core/src/datagen/mod.rs

```diff
@@ -0,0 +1,40 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module contains various utilities required to generate ICU4X data files, typically
+//! via the `icu4x-datagen` tool. End users should not need to consume anything in this module
+//! as a library unless defining new types that integrate with `icu4x-datagen`.
+//!
+//! This module can be enabled with the `datagen` feature on `icu4x-provider-core`.
+
+mod data_conversion;
+mod heap_measure;
+
+pub use data_conversion::{DataConverter, ReturnedPayloadError};
+pub use heap_measure::{HeapStats, HeapStatsMarker};
+
+use crate::any::AnyMarker;
+use crate::buf::BufferMarker;
+use crate::iter::IterableDynProvider;
+use crate::serde::SerializeMarker;
+use crate::DataMarker;
+
+/// A blanket-implemented trait that allows for all the datagen-relevant traits
+/// to be used in trait objects. The type parameter is the marker type needed for
+/// [`IterableDynProvider`].
+pub trait OmnibusDatagenProvider<M: DataMarker>:
```

Comment:
Can we just call this `DatagenProvider`?

Actually in #1092 we said that `IterableDynProvider` is not the right trait for runtime, so maybe we can just rename that `DatagenProvider` and make it extend the `DataConverter` impls? It should also be behind the `datagen` feature now that that exists.

Reply:
I'd rather do wider refactors around what's under what feature as a result of https://github.com/unicode-org/icu4x/issues/1320 (which I plan to get done soon)

Reply:
Still, what do buses have to do with this

Reply:
"Omnibus" is just a term used in English to talk about things that contain more things than usual, typically in the case of media around anthologies and "3-in-1 books".

In this case, this is really just "a lot of traits in one trait", hence the trait name.

Reply:
I have never seen it used that way, maybe it's an American thing. I'd prefer if we avoid such words, we should keep our code accessible to non-Americans and non-native English speakers.

Reply:
Hmm, so I kinda prefer the longer name because we have a lot of FooProvider types already and it gets confusing, i think it would be useful to have a longer name here which is why I picked Omnibus. I could use AllInOne but every time I see AllInOne I just feel like Omnibus is a better name there.

Reply:
Does this need to be a public type, or could it be internal to the datagen crate? As long as everything outside of datagen implements the "actual" traits, then we can safely add the blanket impl in datagen, right?

Reply:
No, because the cldr/uprops/etc providers need to make a Multi provider of trait objects.

Reply:
I thought we could implement a trait on a foreign type so long as that impl is in the same crate as the definition of the trait. We could implement Omnibus for Multi inside of the datagen crate, no?

Reply:
@sffc we could; the problem is the trait objects involved when constructing the Multi in provider_cldr (etc)

the trait needs to exist then, otherwise the fact that it is implemented on each subprovider gets erased

Reply:
I see. I didn't realize we had a trait object at the library boundary. This is a recent change from @robertbastian's refactoring work.

I think a better design is to go back to concrete types at the library boundary.

To be clear, I think we can/should do this in a follow-up PR.

Reply:
I'm unsure how we can do concrete types at the boundary with MultiBlob unless we do some kind of tuple implementation (since Rust doesn't have variadic generics)

We actually don't need this trait at all if not for this boundary :smile:, it purely exists to make that trait object work.

Reply:
Previously we exported each provider on its own (NumbersProvider, TimeZonesProvider, CalendarsProvider, ...). Maybe we should go back to that model.

Reply:
Ah, gotcha. It _is_ convenient to be able to have the list of all sub-providers in one place though

Reply:
(perhaps we could have that list be in the datagen _crate_)

Reply:
> (perhaps we could have that list be in the datagen _crate_)

This would be fine with me. Alternatively: https://github.com/unicode-org/icu4x/pull/1684

Reply:
That is a cool solution, can you incorporate #1684 ?

Reply:
It's not sufficient in and of itself without further work: you either need to name the really complex type over the boundary or you need to do a bunch of manual passthrough implementations. I'd rather land this first, I don't consider this additional trait to be that much overhead.